### ISSUE TITLE
[css-color-6] color-contrast() should allow specifying multiple contrast algorithms that need to be satisfied

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#7357]
Assuming we can specify the contrast algorithm at all (see https://github.com/w3c/csswg-drafts/issues/7356), we should be able to specify *multiple* of them as a safeguard for algorithm bugs, as well as to satisfy legal constraints.

E.g. we know that [WCAG 2.1 contrast is severely broken](https://www.cedc.tools/article.html), yet it is legally mandated that websites pass it. Once we have a better contrast algorithm, we may want `color-contrast()` to find us a color pair that **both** satisfies WCAG 2.1, **as well as** the new improved algorithm.

Syntax could just be a space-separated list of contrast algorithms.

*(Issue filed following breakout discussions between @svgeesus, @fantasai, @argyleink and myself)*

**Comment by argyleink** [comment]
`color-contrast(white vs yellow, blue to wcag3(AA) hct(40) wcag2(AA))` is starting to be a mouthful, but could be mitigated with custom properties and supports queries. 

```css
@supports (color: color-contrast(#F00 vs #00F, #0F0 to wcag3(AA))) {
  :root {
    --target-ratio: wcag3(AA);
  }
}

.custom-label {
  background: var(--some-bg);
  color: color-contrast(var(--some-bg) vs #111, #eee to var(--target-ratio, wcag2(AA)));
}
```

**Comment by alastc** [comment]
> E.g. we know that [WCAG 2.1 contrast is severely broken](https://www.cedc.tools/article.html),

I think that is overblown, the article says (several times) "if the ‘APCA Lightness Contrast’ is more accurate...". If it isn't, the conclusions are not applicable. 

When I've run usability testing with people with low vision, there has been good correlation between colours they struggled with, and WCAG 2 fails. 

I've been following the work (as a non-colour expert) and I think APCA is probably a better formula, and we should continue to work incorporating a better formula into WCAG 3.

I'm just requesting that people in W3C don't use language like "severely broken", when the overall impact of the current guideline is still a net positive.

**Comment by svgeesus** [comment]
Also from that article:

>  the range of possible background colours for black text reduces by approximately 47% (for these particular thresholds).

> the range of possible background colours for white text increases by approximately 63% (for these particular thresholds). 

47% false positives and 63% false negatives with a sample size of 8000 tested color pairs does not fill one with confidence. "Broken" seems applicable.

> "if the ‘APCA Lightness Contrast’ is more accurate..."

In the cited article, on a calibrated screen and with my normal color vision (slight age-related macular yellowing) the APCA was more accurate in all cases.

I agree that more testing is needed.

**Comment by Myndex** [comment]
Hi Lea @LeaVerou 

> .... _as well as to satisfy legal constraints....WCAG 2.1 contrast ... is legally mandated that websites pass it._

Legally mandated is a strong term for the narrow areas where there is actual codification into law.

In the USA, the ADA does not mandate WCAG 2 contrast (indeed, the native ADA signage regulations were gutted of any specific contrast guidance regarding architectural signage some time ago). For government sites and government procurement, the 508 rules **_do_** specify WCAG 2 contrast, but with **two big exception clauses:**

**1. Commercially available.** If something is needed but no commercially available solution is WCAG 2 compliant, it does not have to comply.
**2. Alternate method.** An alternate method can be used so long as it provides for equivalent or better accessibility.

As for case law, the 11th circuit vacated Winn-Dixie in February, so that is moot.

Above is federal level. For state level there is _mainly_ New York and California. I don't have Lexus Nexus access at the moment, but from what Ive seen, no cases relating to contrast have gone to trial and won on merits. Most are out of court settlements, and I'm going to guess many of these _were_ relying on Winn-Dixie.

For other nations it's a grab bag, but in nearly all cases the specification of WCAG 2 is limited to governmentally controlled entities or sites. An exception is Finland. In Australia it extends to non-governmental sites, but last I checked it was level A only, so contrast is not included. In Canada, there is some case law, but a number of exclusions.


> _....we may want color-contrast() to find us a color pair that both satisfies WCAG 2.1, as well as the new improved algorithm._

I'd like to introduce you to Bridge-PCA. It is fully backwards compatible to WCAG 2 but using APCA technology, it fixes the problem of false passes. What is lost is the greater design flexibility of the full APCA guidelines. Bridge-PCA was created specifically to answer the question of "meeting legal obligations by the absolute letter of WCAG 2, regardless of actual veracity".

I do not suggest Bridge PCA as a permanent solution—it is specifically a stop-gap, stepping-stone to address various concerns. It does a much better job calculating for dark mode for instance, and also has enhanced conformance levels.  The npm package is:

    npm i bridge-pca

And the demo tool is https://www.myndex.com/BPCA/

**Comment by Myndex** [comment]
> .... _HCT which uses the L axis from CIE Lab (Tone)_ ....
> _Note that the minimum contrasts for small and large text (50 and 40) are different from the thresholds for WCAG 2.1 and for APCA; thresholds are algorithm-specific._

I've been watching James' developments with interest. I'm a little surprised at the use of L* instead of J.

∆L* has an interesting attribute in that is sort of lines up with WCAG 2 — the implication is, that contrast using a plain L* suffers the same issues as WCAG2.

Digging up some of the early comparison tables, here's one with ∆L*:

<img width="440" alt="Simple WCAG Weber ModWeb Lstar Mich SAPC98G" src="https://user-images.githubusercontent.com/42009457/174482102-e83d35d0-a251-4e90-9e51-46032e2bcfed.png">

An L* simple difference, when the lightest color is white, lines up with a little less than 40 = WCAG 3:!, 50 = 4.5:1, and 62 = 7:1 ... and still as colors get darker, contrast is over reported. I've also tried this with multiple offsets, but that still is not polarity sensitive. 


### But wait, isn't L* perceptually uniform?

L* is based on Munsell value, so it _is_ perceptually based.... _on diffuse reflected light using large color patches, when the observer is in the defined ambient/adaptation conditions._

Viewing **_text_** on a self-illuminated monitor is a different matter, and a different perception. "Middle contrast" for text and other high spatial frequency stimuli isn't at 18%, it's up between 34 and 42 (ish).

As such, ∆L* without some offsets and massaging is not much different than WCAG 2 as this chart shows:

<img width="509" alt="WCGACLstarSAPC98Gat60" src="https://user-images.githubusercontent.com/42009457/174482738-b9348aca-d83f-46d1-a9e7-75211063dfa7.png">

**Comment by Myndex** [comment]
> _I agree that more testing is needed._

Me too. And I realize I need to start publishing sooner rather than later.

**Comment by svgeesus** [comment]
> I've been watching James' developments with interest. I'm a little surprised at the use of L* instead of J.

Yes. Going to all the trouble of taking viewing conditions into account, to calculate the hue and chroma from CIECAM16, while ignoring them to calculate tone (L*), is odd, and not well explained in the article.

**Comment by svgeesus** [comment]
What is SmrsModWbr? I assume it is a modified Weber, got a reference?

**Comment by Myndex** [comment]
> What is SmrsModWbr? I assume it is a modified Weber, got a reference?

"Somers Modified Weber" is a further offset and scaling following an idea from Peli/Hwang's Modifed Weber, and I indicated that was from a series of evaluations in 2019 of various contrast maths. I abandoned it as it does not track the full range very well, and reverse polarity is also sketchy.

The thing with Weber and Michelson contrasts is that they track at very low, threshold contrasts, but they do not predict what happens at supra-threshold readability levels, and the difference is significant.

Maureen Stone (PARC, NIST) and Larry Arend (NASA) had written about using ∆L* for luminance contrast, and in some experiments adding in scaling and offsets, that avenue began to indicate the shape of the perception. This led to greater consideration of CAMs and perceptual models, notably CAM02, R-Lab, Barten's model and Hunt's model. When you consider that viewing a self-illuminated display and reducing to luminance (as that is the parameter essential for readability) this allows a subset of CAM input conditions, permitting a reasonable simplification to determine luminance & stimulus size-based readability contrast via perceptual lightness/darkness difference.

**Comment by jpohhhh** [comment]
Here's a Google Sheet for visualizing this. You should be able to edit contrast ratio/SAPC values and see the graphs updated
 https://docs.google.com/spreadsheets/d/1G6rZInfua3Y8125avr2_OczDfZ7eWdqhFywRpvbdS1U/edit?usp=sharing

**Comment by Myndex** [comment]
Hi @jpohhhh

This material is so much easier to discuss live with examples, instead of text, but here's a bunch ahead of the call.

First, I do want to mention as _kindly as I possibly can_ that your take on APCA does not reflect the reality or underlying theory. So I want to ask if you've read any of the documentation or white papers? My concern is that I must not have explained things correctly, which is apparently my Achilles heel !! The canonical documentation is at the main repo, and I'll list links in order of preferred reading later in this post.

## _But very quickly_
James, I am wondering where you read or came to the following opinion:
> _...and frankly, the fact APCA neither explicitly models flare, nor does its delta L behavior show it reflecting that white cant get lighter but black can lighter in the presence of flare..._

### _I ask because these statements are categorically false....!_ 😳
- APCA **_EXPLICITLY_** models flare, and does so in a way similar to how DICOM does, and this is documented several places, and is evident in the math.
- APCA also directly addresses the asymmetric nature of polarity, _AND_ models, and more importantly **compensates** for, the _VARIETY_ of psychophysical phenomena that occur near black. Even the simple-pair version of the math and/or the guidelines address these directly, Including and not limited to flare, rod intrusion, halation, and polarity.

I am very concerned that such a line of thought is out there—did you read this somewhere? What prompted this?


## Further Addressing Your Posts Above:

> _APCA as documented for WCAG 3 draft_

Which draft?  I looked at the spreadsheet—but the functions are hidden so I can't examine the math or method....?? It does not look right. See the readme at the main repo, or at the apca-w3 npm page. **_Do not use anything labeled SAPC._**

### One or two little bits:

> _Contrast measures are for a11y, guaranteeing legibility for the population._

**Contrast is important for 100% of sighted users.** Human vision is a spectrum as wide as the human experience, and our vision changes over our lives: we're essentially blind at birth, and it takes 20 years to develop peak contrast sensitivity... and then we hit our 40s, and presbyopia sets in. At 60, it's all down hill from there!

> _how you're handling the ~10% of users with skews in perception of hue and chroma (some can't see it entirely!)_ 

The short answer is that the APCA guidelines are directly following the long established, peer reviewed scientific consensus of modern readability research, especially for low vision, and particularly Dr. Lovie-Kitchin, Bailey, Whittaker, et alia, and also Legge, Chung, etc.....

Okay, we need to define a couple things:

**1) Readability Contrast**—Prime Focus of APCA
    - For the vast majority of text, and particularly for columns of body text, only achromatic luminance contrast is useful or important, especially for low vision.
    - Hue/Chroma are processed in the brain separately, and don't figure into lexical processing (for the most part, see "Stroop tests" for more).
    - **ALL sighted users need ample luminance contrast for readability** at best speed and comprehension.

**2) Discernibility Contrast**—A Different Beast, processed differently in the brain.
    - Here's the problem, colors **are** very helpful for object recognition. Dataviz fits in here of course.
    - It is THESE tasks that those with color insensitive vision need help with, and that help is summarized as:
        - **Do not use color as the sole means of conveying information**

3) **The Term "Contrast":**
    - Like color, contrast is not real. Both color and contrast are **_perceptions_** of the Human Vision System (HVS).
    - Luminance is a photometric _QUANTITY_ of light. It is not "perception".
    - A luminance difference is a photometric distance between the light and dark values.
        - **This is not contrast.**
    - Contrast, the perception, is a psychophysical sensation
    - Contrast is very context dependent, and some of the contextual factors are:
        - Spatial Frequency of the stimuli, meaning font weight, line thickness, size, spacing, edge sharpness....
        - Adaptation: local, field, and ambient (global) adaptation state
        - The stimulus luminance in a multi-way race with the adjacent (local) luminance, and considering the field luminance.
        - **The distance above threshold (JND).**
 
This last point is the most critical to understand, because not only are perceptual lightness estimation power curves curved, but the perception of contrast resulting from the distance of two different lightnesses is ALSO curved, by which I mean the contrast change from threshold to supra-threshold is also very non linear relative to the distance (difference). And the shape of THAT curve is ALSO dependent on spatial frequency.

Take a look at this image:

![Yellow Dots checkers Optical_grey_squares_orange_brown_(opaque)](https://user-images.githubusercontent.com/42009457/178430040-7f41c09c-fbdf-4b39-8136-8210cc34bc1d.png)

The two yellow dots are EXACTLY the same. As far as XYZ or LAB or the sRGB vaules being sent to the monitor are concerned, both yellow dots are identical.

And yet they both look distinctly different. 


The photometric difference between two light or dark "things" **does not define contrast perception**. It is _not_ contrast.

And neither WCAG 2 ratio or ∆L* are uniform to perception. Historically this has not been a problem BECAUSE, in physical PRINT, it is almost always black ink on white paper. And even if not, there was a designer there to lock it in place.

### _Why This Is So Important Today_
The WEB is dynamic content, not locked into place as printed words on paper are, and today there is the desire to have **automatic properties in CSS** and automation for things like auto darkmode or auto high-contrast.

**AUTOMATION of colors is only realistically possible if you have perceptually uniform methods.**

While L* may be more or less uniform to perception of low spatial frequency diffuse surfaces under ideal illumination conditions, I can tell you that L* does not predict lightness perception of high spatial frequency elements (text) on a self illuminated monitor.

I'll fill you in on this this afternoon, but in short:

### Human perception is very literally curves upon curves.

**Now a few things to set the record straight, particularly for anyone reading at home....**

-----
## _RE: APCA_
1. APCA does not use L\*, which is based on Munsell value, which is lightness perception of patches of surface colors while adapted to a viewing environment with a defined ambient illumination.
2. If anything, APCA is closer to Stevens, but more specifically references appearance models including RLab, Hunt, Barten, and CAM02, among others.
3. APCA **_does_** have a mild hue adjustment, and it's part of the "non-standard" linearization, which is also documented.[footnote a] 

## _COLOR VISION INSENSITIVITY_
There is this misunderstanding out on the web, and I am not sure the source, but I do see it a lot in the context of accessibility. I believe it may be in conjunction with the WCAG 2.0 understandings documents. Here are the facts.

1. **It is peer reviewed scientific consensus that individuals with color vision deficiency have standard _or better_ contrast sensitivity and visual function.**
    - Peer reviewed literature: [1][2][3]
    - The one exception is red against black for protan individuals, as they have reduced luminance perception of deep reds.
    - Those with deutan forms are not adversely affected in terms of contrast sensitivity, as the M cone is substantially overlapped by the L cone.
    - In fact deutan/protan show standard **_or better_** visual function _overall_ (hue discrimination notwithstanding).
        - A current theory suggests that the deutan/protan improved acuity is the result of lower neural noise.
    - It is however known that a general loss in contrast sensitivity may cause a loss in color differentiation as a co-morbid affect[4][5], which one might expect due to the lower contrast resolution of the color pathways.

2. It is peer reviewed scientific consensus that achromatic luminance contrast is what is most important for readability, and also for fine details.[6][7]
    - Hue/chroma do not enter into readability and high spatial frequency stimuli, and especially for low vision.
        - Hue chroma are a third the resolution and a third the contrast strength of the luminance channel.
    - The exception is for the few instances where hue is a blocking condition. Examples:
        - Red against Black
            - Red too low of a luminance for small text, even for normal vision.
            - But for the protan it is essentially unreadable,[8] and
            - With some of the newer technology UHD monitors, the red could be invisible to the protan. 
        - Red against Blue
            - For most vision types, red against blue can result in chromatic aberration, with the symptoms of "shimmer" or "chromostereopsis".[9]
        
-----
## _APCA DOCUMENTATION_
I've been working hard to clear up these misconceptions, so I've been trying to organize the documentation is a logical manner. The links listed in this section are placed in an order that starts with the plain language overview before getting into the minutiae. 

### Easy Start &darr;
- [Why APCA](https://git.apcacontrast.com/documentation/WhyAPCA) The brief overview, starts with why WCAG2 has issues, discusses the challenges of readability contrast, and then introduces the solution.
- [Regarding Exponents](https://git.apcacontrast.com/documentation/regardingexponents) The brief discussion of the use of the exponents in APCA, including why 2.4 is NOT a mistake, but an intentional pre-shaping stage.
- [Main ReadMe Doc](https://git.apcacontrast.com/documentation/README) In the main repo documentation folder, this includes a code walkthrough, LaTeX math, the optional visual font lookups, and links to more stuff.

### Visual Comparisons and Related Articles &darr;
- [Orange You Wondering About Contrast?](https://gist.github.com/Myndex/1dadb6dcac596f1cd7a5686a076f697f) Part I of three on the APCA and why perceptually uniform lightness contrast is important.
- [Better reading on the web](https://uxdesign.cc/better-reading-on-the-web-c943c4cfc91a) A discussion of automated testing and real world examples of contrast metrics.
- [Please Stop Using Grey Text](https://tangledweb.xyz/please-stop-using-grey-text-3d3e71acfca8) Discusses issues with WCAG2, with examples of dark mode, etc.
- [WCAG2 vs APCA](https://github.com/Myndex/SAPC-APCA/discussions/30#discussion-3678164) A set of direct head to head comparisons, including illustrative graphs and apples to apples comparisons studies.
- [A Contrast of Contrasts](https://tangledweb.xyz/a-contrast-of-contrasts-3ddda0f4061b) Brief article on perceptual contrast.

### The Official Demo Tool  &darr;
- [The APCA Demo Tool](https://www.myndex.com/APCA/) This is the canonical example tool for the current APCA.

### Deeper Technical & Theory  &darr;
These are all a bit rough or in draft form, and are much deeper dives into the underlying theories.
- [Experiment CE17, ](https://www.myndex.com/WEB/WCAG_CE17polarity) a short early white paper on the concept.
- [The white paper in progress](https://www.w3.org/WAI/GL/task-forces/silver/wiki/Visual_Contrast_of_Text_Subgroup/Whitepaper) regarding the underlying theories of readability contrast, with a partial bibliography.
- [Practical Contrast Guidelines](https://github.com/Myndex/SAPC-APCA/discussions/39#discussion-3757102) work in progress in defining use case guidelines and how they inform contrast design decisions, includes more theory and referenced research in readability.

-----


That's it for this post, thank you for reading.

—_Andy_

----
----
### _Academic Peer Reviewed Published Science as referenced above_
[1 • Spatial visual function in anomalous trichromats: Is less more?](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0209662)
[2 • Eye-Tracker Analysis of the Contrast Sensitivity of Anomalous and Normal Trichromats: A Loglinear Examination with Landolt-C Figures](https://www.mdpi.com/2076-3417/11/7/3200/htm#)
[3 • Contrast sensitivity of patients with congenital color vision deficiency](https://metrovision.fr/2019/2019_Ilhan_Contrast_sensitivity_in_congenital_color_deficiency.pdf)
[4 • Evaluation of contrast sensitivity and color vision in lead and zinc mine workers](https://pdfs.semanticscholar.org/c44b/3e2ca75ce8247bbb2eb6126db99542f366b3.pdf)
[5 • Effects of Contrast Sensitivity on Colour Vision Testing](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5762144/)
[6 • Effects of luminance contrast and character size on reading speed](https://www.sciencedirect.com/science/article/pii/S0042698919302111)
[7 • Luminance and chromatic contrast effects on reading and object recognition in low vision](https://pubmed.ncbi.nlm.nih.gov/8594527/)

### Non-Academic General Audience
[8 • What’s Red & Black & Also Not Read?](https://tangledweb.xyz/whats-red-black-also-not-read-573b9c0a97ed)
[9 • Chromostereopsis](https://en.wikipedia.org/wiki/Chromostereopsis)

_Footnotes_       
(a) Nevertheless, I am removing non-standard linearization, and using the standard(s) per CSS 4/5, and handling protan compensation with a separate protan color module shortly...
- I would have used ^2.5 if I had realized the massive misunderstandings that would happen by using ^2.4 without the offset (i.e. the piecewise), even though I documented it...
    - Related, it's been brought to my attention my documentation is hard to parse, and I am actively working to resolve that.

**Comment by Myndex** [comment]
Hi @jpohhhh 

> 2. ..._to the main point that it doesn't make sense to describe WCAG 2.0 as severely broken, that's very charged language, and _from your perspective_, I agree. having that perspective is what motivates one to work on problems like this._

Just to point out, this is not new (I started this project with post #695 circa April 2019) and I am not the only one, this issue has been widely criticized, including back in 2007 when objections from IBM were ignored for instance. I layout the basis of the problems in the 44,000 word thread #695 back in 2019.

> _However, once other people, in a more formal setting, justify that perspective via a couple people with standard color vision noticing that magnitude of luminance difference is different from magnitude of color difference...that's...not good. At all._

First, this is not true on the fact of it. I do not have standard vision. I WAS legally blind due to severe early onset cataracts, and now 6 surgeries later I have low vision. Yay. 

But the conflation of visual function and color insensitivity is a spurious one. For readability, only achromatic luminance contrast is critical. Color is useful for discrimination of objects, but not for reading. These are two completely separate visual functions.

> 4. _I know about red/black, WCAG 2.x has called that out explicitly._

It is only mentioned in the "understanding document" but is NOT considered in the algorithm. In APCA, the algorithm specifically derates red and fails it as part of the math. And the protan compensator does so even more strongly.

**Comment by svgeesus** [comment]
>     The contrast algorithms we've seen, whether WCAG 2.1 or APCA, take two Ys as input

Not exactly. WCAG 2.1 takes two luminances (CIE Y) as input. APCA uses a non-standard transfer function of linearizing sRGB and thus the `Ys` term is not the same as CIE Y, and cannot be computed from it. It requires the individual non-linear color component values as starting values. I [reported that here](https://github.com/w3c/silver/issues/643) and the conclusion was that this is intentional, nt an approximation error.

>    Y and L* are the same physical quantity. 

No.

> They are pure functions of each other. They can be converted to and from each other with no other inputs.

Yes.

>    Thus, given a L* and contrast measure, we can find the L* that contrasts

Yes. We convert L* to Y, compute the other Y, and then can if we want convert that back to L*

>    Given that, the L* and the L* needed for contrast, we can calculate delta L* required for contrast

For _that color_, yes, _if_ we want to replicate the WCAG 2.1 algorithm. Doing so seems pointless because the problem with WCAG 2.1 is that is uses a non perceptually-uniform measure (CIE Luminance) instead of a perceptually uniform measure (CIE Lightness).

**Comment by svgeesus** [comment]
In preparation for discussing the various contrast algorithms, the [color.js documentation on contrast might be useful](https://colorjs.io/docs/contrast.html) may be helpful.

_Note_: Weber and Michelson are broken in color.js currently, investigating.

**Comment by fantasai** [comment]
We resolved to add `<contrast-algo>+` to `contrast-color()` where `<contrast-algo>` represents a contrast function.

**Comment by svgeesus** [comment]
> Note: Weber and Michelson are broken in color.js currently, investigating.

I fixed that bug, which was a stupid typo of = for === /facepalm

**Comment by LeaVerou** [comment]
> > Note: Weber and Michelson are broken in color.js currently, investigating.
> 
> I fixed that bug, which was a stupid typo of = for === /facepalm

They still seem broken [here](https://colorjs.io/apps/blackwhite/), unless they really are _that_ bad.

**Comment by svgeesus** [comment]
By broken I mean that they were returning `+Inf` as the contrast ratio, regardless of inputs. A test that was intended to trap division by zero actually assigned zero to the denominator :)

Yes, they are both not very good, especially for light text on dark backgrounds.

**Comment by NateBaldwinDesign** [comment]
I'm curious to know, is there going to be specific formulas that are supported by `color-contrast()`? Will it only support WCAG 2.x relative luminance and APCA, or are there plans to support more (and if so, which ones and why)? This is based on some questions and concerns I [mentioned here](https://github.com/w3c/csswg-drafts/issues/7359#issuecomment-1209578115).

**Comment by Myndex** [comment]
Hi @LeaVerou and @svgeesus 

> _...unless they really are that bad._

In the context of the black/white page, both the unmodified Weber and Michelson have a polarity sensitive issue, as neither are perceptually uniform. While both are useful in research relating to the JND, neither are useful for practical design guidance for supra-threshold contrasts.

I evaluated these and all other available contrast models in 2019, along with many variants (some of which I mention later in this post).

As for the black and white flip page:

```js
Weber:     (lighter - darker) / darker
Michelson: (lighter - darker) / (lighter + darker)
```

So if white is 1 and black is 0, we can see why both of these fail to define a useful "flip point".

### For White:
```js
Weber:     (1.0 - color) / color
Michelson: (1.0 - color) / (1.0 + color)
```

### For Black:
```js
Weber:     (color - 0) / 0
Michelson: (color - 0) / (color + 0)
```

So, as we can see for black, Weber produces infinity and Michelson = 1, and in both cases white vs any color will never be the higher compared to black vs any color.

### Regarding the infinity fix:

in Weber.js, there is:
```js
    return Y2 === 0 ? 0 : (Y1 - Y2) / Y2;
```

To fix devide by 0, which would be infinity. But in returning 0, it hides that the actual result should be a _maximum_. As a result, in the black/white page, the Weber shows white text, when in reality it should show black text for all cases similar to Michelson, due to the nature of these algorithms.

If I may suggest to consider instead:

```js
    return Y2 === 0 ? 50000 : (Y1 - Y2) / Y2;
```
The reason: the darkest sRGB color above black is `#000001` and this produces a plain Weber contrast of ~45647. So, setting the divide-by-zero result at 50000 is a reasonable max clamp for the plain Weber.

-----
## Bonus Round!

I don't know if you want to play with these, but there are other variants, some are interesting, and we evaluated all of them in 2019. Among the variants is a couple modified Webers where a brute-forced offset is added to the denominator. Sometimes this is claimed to be a "flare" component, but in reality is in effect a "push" to a suprathreshold level.

### _Three Mod Webers_
These assume Y is 0.0 to 1.0:

```js
    hwangPeli = (Y1 - Y2) / (Y2 + 0.05);

    somersB = ((Y1 - Y2) / (Y2 + 0.1)) * 0.9;
    somersE = (Y1 - Y2) / (Y2 + 0.35);
```

However these do not track polarity changes particularly well, and have a mid-range "bump".

### 𝜟𝜱✴︎
(delta phi star)    

A better and interesting modification is this delta L* variant we created on the path toward SACAM (and APCA).

Here, create Lstar from the piecewise sRGB->Y and L* per the standard CIE math, then:

```js
    deltaPhiStar = (bgLstar^1.618 - txLstar^1.618)^0.618;
```

This mainly works for "Light Mode" but does not track dark mode quite as well. Also, while this is close to parity with light mode APCA at Lc +90, lower contrasts are over-reported, and it does not match in dark mode. Some of this can be addressed with scales and offsets.

Nevertheless, I thought you might find these variants interesting.

APCA builds on these early experiments, but has added value in terms of polarity sensitivity and wider range for better guideline thresholds.


### _Double Bonus Round_
Regarding the simple concept of a black/white flip, I have this interactive demo-page:  [ **_FlipForColor_** ](https://www.myndex.com/WEB/FlipForColor) which includes a brief discussion.

For a deeer dive, there is a [CodePen](https://codepen.io/myndex/pen/GRyxrrm), and a [Repo](https://github.com/Myndex/fancyfontflipping), and a [Gist](https://gist.github.com/Myndex/e1025706436736166561d339fd667493#lets-flip-for-color) that discusses this and related issues including font size and weight as it relates to flipping.



Thank you for reading

**Comment by LeaVerou** [comment]
> ### Regarding the infinity fix:
> in Weber.js, there is:
> 
> ```js
>     return Y2 === 0 ? 0 : (Y1 - Y2) / Y2;
> ```
> 
> To fix devide by 0, which would be infinity. But in returning 0, it hides that the actual result should be a _maximum_. As a result, in the black/white page, the Weber shows white text, when in reality it should show black text for all cases similar to Michelson, due to the nature of these algorithms.
> 
> If I may suggest to consider instead:
> 
> ```js
>     return Y2 === 0 ? 50000 : (Y1 - Y2) / Y2;
> ```
> 
> The reason: the darkest sRGB color above black is `#000001` and this produces a plain Weber contrast of ~45647. So, setting the divide-by-zero result at 50000 is a reasonable max clamp for the plain Weber.
> 

Pull request?

**Comment by svgeesus** [comment]
> Pull request?

Oh, that would have been a good idea but [already done](https://github.com/LeaVerou/color.js/commit/a078113e6bf4122fa0fd037a1cf4866bc029fba2)

**Comment by LeaVerou** [comment]
Closing this, as @fantasai and I made the prose edits, however we currently only have one algorithm.
